sktree/tree/_unsup_criterion.pyx

```diff
@@ -136,35 +164,165 @@ cdef class TwoMeans(UnsupervisedCriterion):
-        pass
+        """Evaluate the impurity of the current node.
+
+        Evaluate the TwoMeans criterion as impurity of the current node,
+        i.e. the impurity of sample_indices[start:end]. The smaller the impurity the
+        better.
+        """
+        cdef double impurity
+
+        impurity = self._compute_variance(self.start, self.end) / self.weighted_n_node_samples
+
+        return impurity
+
+    cdef double _compute_variance(self, start, end) nogil:
+        """Computes variance of left and right cluster at `pos`
+        """
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef DOUBLE_t w = 1.0
+        cdef SIZE_t n_samples = end - start
+
+        cdef double var = 0.0 #variance
+        cdef double mean = 0.0 #mean
+
+        cdef const DTYPE_t[:,:] X = self.X
+        
+        #calculate the mean
+        for p in range(start, end):
+            i = self.sample_indices[p]
+
+            mean += self.X[i] / n_samples
+
+        #calculate variance
+        for p in range(start, end):
+            i = self.sample_indices[p]
+
+            if self.sample_weight is not None:
+                w = self.sample_weight[i]
+
+            var += w * (X[i] - mean) * (X[i] - mean)
+
+        return var / n_samples
 
     cdef void children_impurity(self, double* impurity_left,
-                                double* impurity_right) nogil:
-        # need to implement
-        pass
+                                    double* impurity_right) nogil:
+       """Evaluate the impurity in children nodes.
+
+        i.e. the impurity of the left child (samples[start:pos]) and the
+        impurity the right child (samples[pos:end]).
+
+        where the impurity of two children nodes are essentially the variances of two nodes
+
+        left_variance = left_weight * left_impurity / n_sample_of_left_child
+        right_variance = right_weight * right_impurity / n_sample_of_right_child
+
+        Parameters
+        ----------
+        impurity_left : double pointer
+            The memory address to save the impurity of the left node
+        impurity_right : double pointer
+            The memory address to save the impurity of the right node
+        """
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t start = self.start
+        cdef SIZE_t end = self.end
+
+        impurity_left = _compute_variance(start,pos) / self.weighted_n_left
+        impurity_right = _compute_variance(pos,end) / self.weighted_n_right
 
     cdef void node_value(self, double* dest) nogil:
-        """Compute the node value of samples[start:end] and save it into dest.
+        r"""Set the node value with sum_total and save it into dest.
 
         Parameters
         ----------
         dest : double pointer
             The memory address which we will save the node value into.
         """
-        # need to implement
-        pass
+        # memcpy(dest, &self.sum_total, sizeof(double))
+        dest = self.sum_total
```

Comment:
```suggestion
        # assign value of `sum_total` to the memory pointed to by `dest`
        *dest = self.sum_total
```

I would recommend reading up on C++ pointers to just get an idea for what this is doing